File: apps/api/coinmark_api/ingest/orderbook_aggregator.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass
from decimal import Decimal

from coinmark_api.ingest.buckets import bucket_ms, floor_bucket_start_ms


@dataclass(frozen=True, slots=True)
class OrderbookBucketKey:
    market: str
    symbol: str
    bucket: str
    bucket_start_ms: int


@dataclass(slots=True)
class OrderbookBucketDelta:
    spread_bps_sum: Decimal = Decimal("0")
    depth_imbalance_l5_sum: Decimal = Decimal("0")
    microprice_shift_bps_sum: Decimal = Decimal("0")
    wall_pressure_l5_sum: Decimal = Decimal("0")
    sample_count: int = 0

    taker_buy_notional: Decimal = Decimal("0")
    taker_sell_notional: Decimal = Decimal("0")

    depletion_events: int = 0
    replenishment_events: int = 0

    l1_depth_baseline: Decimal | None = None
    depletion_active: bool = False

    def add_orderbook_sample(
        self,
        *,
        spread_bps: Decimal,
        depth_imbalance_l5: Decimal,
        microprice_shift_bps: Decimal,
        wall_pressure_l5: Decimal,
        l1_depth_notional: Decimal,
    ) -> None:
        self.spread_bps_sum += spread_bps
        self.depth_imbalance_l5_sum += depth_imbalance_l5
        self.microprice_shift_bps_sum += microprice_shift_bps
        self.wall_pressure_l5_sum += wall_pressure_l5
        self.sample_count += 1

        if self.l1_depth_baseline is None:
            self.l1_depth_baseline = l1_depth_notional
            return

        baseline = self.l1_depth_baseline
        if baseline <= 0:
            self.l1_depth_baseline = l1_depth_notional
            return

        drop_threshold = baseline * Decimal("0.60")
        recover_threshold = baseline * Decimal("0.85")

        if (not self.depletion_active) and l1_depth_notional < drop_threshold:
            self.depletion_events += 1
            self.depletion_active = True
        elif self.depletion_active and l1_depth_notional >= recover_threshold:
            self.replenishment_events += 1
            self.depletion_active = False

        self.l1_depth_baseline = baseline * Decimal("0.98") + l1_depth_notional * Decimal("0.02")

    def add_trade(self, *, taker_buy_notional: Decimal, taker_sell_notional: Decimal) -> None:
        self.taker_buy_notional += taker_buy_notional
        self.taker_sell_notional += taker_sell_notional


class OrderbookAggregator:
    def __init__(self, bucket: str = "1m") -> None:
        self._bucket = bucket
        self._bucket_size_ms = bucket_ms(bucket)
        self._lock = asyncio.Lock()
        self._deltas: dict[OrderbookBucketKey, OrderbookBucketDelta] = {}

    async def add_orderbook_sample(
        self,
        *,
        market: str,
        symbol: str,
        ts_ms: int,
        spread_bps: Decimal,
        depth_imbalance_l5: Decimal,
        microprice_shift_bps: Decimal,
        wall_pressure_l5: Decimal,
        l1_depth_notional: Decimal,
    ) -> None:
        key = OrderbookBucketKey(
            market=market,
            symbol=symbol,
            bucket=self._bucket,
            bucket_start_ms=floor_bucket_start_ms(ts_ms, self._bucket),
        )
        async with self._lock:
            delta = self._deltas.get(key)
            if delta is None:
                delta = OrderbookBucketDelta()
                self._deltas[key] = delta
            delta.add_orderbook_sample(
                spread_bps=spread_bps,
                depth_imbalance_l5=depth_imbalance_l5,
                microprice_shift_bps=microprice_shift_bps,
                wall_pressure_l5=wall_pressure_l5,
                l1_depth_notional=l1_depth_notional,
            )

    async def add_trade(
        self,
        *,
        market: str,
        symbol: str,
        ts_ms: int,
        taker_buy_notional: Decimal,
        taker_sell_notional: Decimal,
    ) -> None:
        key = OrderbookBucketKey(
            market=market,
            symbol=symbol,
            bucket=self._bucket,
            bucket_start_ms=floor_bucket_start_ms(ts_ms, self._bucket),
        )
        async with self._lock:
            delta = self._deltas.get(key)
            if delta is None:
                delta = OrderbookBucketDelta()
                self._deltas[key] = delta
            delta.add_trade(taker_buy_notional=taker_buy_notional, taker_sell_notional=taker_sell_notional)

    async def drain_closed(self, now_ms: int) -> list[tuple[OrderbookBucketKey, OrderbookBucketDelta]]:
        cutoff = floor_bucket_start_ms(now_ms, self._bucket)
        async with self._lock:
            closed_keys = [k for k in self._deltas.keys() if k.bucket_start_ms < cutoff]
            out: list[tuple[OrderbookBucketKey, OrderbookBucketDelta]] = []
            for key in closed_keys:
                delta = self._deltas.pop(key)
                out.append((key, delta))
            return out

File: apps/ingest-go/internal/ingest/orderbook_aggregator.go
package ingest

import (
	"sync"

	"github.com/shopspring/decimal"
)

type OrderbookBucketKey struct {
	Market        string
	Symbol        string
	Bucket        string
	BucketStartMS int64
}

type OrderbookBucketDelta struct {
	SpreadBPSSum          decimal.Decimal
	DepthImbalanceL5Sum   decimal.Decimal
	MicropriceShiftBPSSum decimal.Decimal
	WallPressureL5Sum     decimal.Decimal
	DepthImbalanceL20Sum  decimal.Decimal
	WallPressureL20Sum    decimal.Decimal
	SampleCount           int64

	TakerBuyNotional  decimal.Decimal
	TakerSellNotional decimal.Decimal

	DepletionEvents     int64
	ReplenishmentEvents int64

	L1DepthBaseline *decimal.Decimal
	DepletionActive bool
}

func NewOrderbookBucketDelta() *OrderbookBucketDelta {
	return &OrderbookBucketDelta{
		SpreadBPSSum:          decimal.Zero,
		DepthImbalanceL5Sum:   decimal.Zero,
		MicropriceShiftBPSSum: decimal.Zero,
		WallPressureL5Sum:     decimal.Zero,
		DepthImbalanceL20Sum:  decimal.Zero,
		WallPressureL20Sum:    decimal.Zero,
		TakerBuyNotional:      decimal.Zero,
		TakerSellNotional:     decimal.Zero,
	}
}

func (d *OrderbookBucketDelta) AddOrderbookSample(spreadBPS, depthImbalanceL5, micropriceShiftBPS, wallPressureL5, l1DepthNotional, depthImbalanceL20, wallPressureL20 decimal.Decimal) {
	d.SpreadBPSSum = d.SpreadBPSSum.Add(spreadBPS)
	d.DepthImbalanceL5Sum = d.DepthImbalanceL5Sum.Add(depthImbalanceL5)
	d.MicropriceShiftBPSSum = d.MicropriceShiftBPSSum.Add(micropriceShiftBPS)
	d.WallPressureL5Sum = d.WallPressureL5Sum.Add(wallPressureL5)
	d.DepthImbalanceL20Sum = d.DepthImbalanceL20Sum.Add(depthImbalanceL20)
	d.WallPressureL20Sum = d.WallPressureL20Sum.Add(wallPressureL20)
	d.SampleCount++

	if d.L1DepthBaseline == nil {
		v := l1DepthNotional
		d.L1DepthBaseline = &v
		return
	}
	baseline := *d.L1DepthBaseline
	if baseline.LessThanOrEqual(decimal.Zero) {
		v := l1DepthNotional
		d.L1DepthBaseline = &v
		return
	}

	dropThreshold := baseline.Mul(decimal.RequireFromString("0.60"))
	recoverThreshold := baseline.Mul(decimal.RequireFromString("0.85"))

	if !d.DepletionActive && l1DepthNotional.LessThan(dropThreshold) {
		d.DepletionEvents++
		d.DepletionActive = true
	} else if d.DepletionActive && l1DepthNotional.GreaterThanOrEqual(recoverThreshold) {
		d.ReplenishmentEvents++
		d.DepletionActive = false
	}

	newBaseline := baseline.Mul(decimal.RequireFromString("0.98")).Add(l1DepthNotional.Mul(decimal.RequireFromString("0.02")))
	d.L1DepthBaseline = &newBaseline
}

func (d *OrderbookBucketDelta) AddTrade(takerBuyNotional, takerSellNotional decimal.Decimal) {
	d.TakerBuyNotional = d.TakerBuyNotional.Add(takerBuyNotional)
	d.TakerSellNotional = d.TakerSellNotional.Add(takerSellNotional)
}

type OrderbookDrainItem struct {
	Key   OrderbookBucketKey
	Delta *OrderbookBucketDelta
}

type OrderbookAggregator struct {
	bucket       string
	bucketSizeMS int64
	mu           sync.Mutex
	deltas       map[OrderbookBucketKey]*OrderbookBucketDelta
}

func NewOrderbookAggregator(bucket string) *OrderbookAggregator {
	size, _ := BucketMS(bucket)
	return &OrderbookAggregator{
		bucket:       bucket,
		bucketSizeMS: size,
		deltas:       make(map[OrderbookBucketKey]*OrderbookBucketDelta),
	}
}

func (a *OrderbookAggregator) ensure(key OrderbookBucketKey) *OrderbookBucketDelta {
	d, ok := a.deltas[key]
	if ok {
		return d
	}
	d = NewOrderbookBucketDelta()
	a.deltas[key] = d
	return d
}

func (a *OrderbookAggregator) AddOrderbookSample(market, symbol string, tsMS int64, spreadBPS, depthImbalanceL5, micropriceShiftBPS, wallPressureL5, l1DepthNotional, depthImbalanceL20, wallPressureL20 decimal.Decimal) {
	startMS, err := FloorBucketStartMS(tsMS, a.bucket)
	if err != nil {
		return
	}
	k := OrderbookBucketKey{Market: market, Symbol: symbol, Bucket: a.bucket, BucketStartMS: startMS}
	a.mu.Lock()
	defer a.mu.Unlock()
	a.ensure(k).AddOrderbookSample(spreadBPS, depthImbalanceL5, micropriceShiftBPS, wallPressureL5, l1DepthNotional, depthImbalanceL20, wallPressureL20)
}

func (a *OrderbookAggregator) AddTrade(market, symbol string, tsMS int64, takerBuyNotional, takerSellNotional decimal.Decimal) {
	startMS, err := FloorBucketStartMS(tsMS, a.bucket)
	if err != nil {
		return
	}
	k := OrderbookBucketKey{Market: market, Symbol: symbol, Bucket: a.bucket, BucketStartMS: startMS}
	a.mu.Lock()
	defer a.mu.Unlock()
	a.ensure(k).AddTrade(takerBuyNotional, takerSellNotional)
}

func (a *OrderbookAggregator) DrainClosed(nowMS int64) []OrderbookDrainItem {
	cutoff, err := FloorBucketStartMS(nowMS, a.bucket)
	if err != nil {
		return nil
	}
	a.mu.Lock()
	defer a.mu.Unlock()

	out := make([]OrderbookDrainItem, 0)
	for key, delta := range a.deltas {
		if key.BucketStartMS < cutoff {
			out = append(out, OrderbookDrainItem{Key: key, Delta: delta})
			delete(a.deltas, key)
		}
	}
	return out
}

func (a *OrderbookAggregator) PendingBuckets() int {
	a.mu.Lock()
	defer a.mu.Unlock()
	return len(a.deltas)
}
